pkg/server/telemetry/middleware.go

```diff
@@ -0,0 +1,79 @@
+package telemetry
+
+import (
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/felixge/httpsnoop"
+	"github.com/gorilla/mux"
+	"github.com/sirupsen/logrus"
+)
+
+type MetricsMiddleware struct {
+	telemetry *Telemetry
+}
+
+func NewMetricsMiddleware(t *Telemetry) *MetricsMiddleware {
+	return &MetricsMiddleware{t}
+}
+
+func (m *MetricsMiddleware) InstrumentHTTPEndpointStats(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+		logger, w := makeMetricsResponseLogger(w)
+
+		route := mux.CurrentRoute(r)
+		path, err := route.GetPathTemplate()
+		if err != nil {
+			logrus.Warn("unable to retrieve path from route for metrics")
+		}
+
+		endpoint := m.ParseRoute(path)
+
+		h.ServeHTTP(w, r)
+
+		if endpoint != "" {
+			m.telemetry.Metrics.IncrCounter([]string{endpoint, Count}, 1)
+			m.telemetry.Metrics.SetGauge([]string{endpoint, Latency}, float32(time.Since(start).Milliseconds()))
+			if logger.Status() >= 500 {
+				m.telemetry.Metrics.IncrCounter([]string{endpoint, "500"}, 1)
+			} else if logger.Status() >= 400 {
+				m.telemetry.Metrics.IncrCounter([]string{endpoint, "400"}, 1)
+			} else if logger.Status() >= 200 && logger.status < 300 {
+				m.telemetry.Metrics.IncrCounter([]string{endpoint, "200"}, 1)
+			}
+		}
+	})
+}
+
+func (m *MetricsMiddleware) ParseRoute(routeName string) string {
+	switch routeName {
+	case "/enroll/{hostname}":
+		return "enroll_host"
+	default:
+		return strings.Trim(routeName, "/")
+	}
+}
+
+type metricsResponseLogger struct {
+	w      http.ResponseWriter
+	status int
+}
+
+func (l *metricsResponseLogger) WriteHeader(code int) {
+	l.status = code
+}
+
+func (l *metricsResponseLogger) Status() int {
+	return l.status
+}
+
+func makeMetricsResponseLogger(w http.ResponseWriter) (*metricsResponseLogger, http.ResponseWriter) {
+	logger := &metricsResponseLogger{w: w, status: http.StatusOK}
```

Comment:
This was a little bit unclear to me, but you init the metricsResponseLogger with 200 and then within `ServeHTTP` `WriteHeader` is called, which will ultimately set the status code that will later be used in the metrics setting? Some comments around this could be helpful.